server/src/test/java/org/elasticsearch/action/DocWriteResponseTests.java

```diff
@@ -96,4 +98,35 @@ public void testToXContentDoesntIncludeForcedRefreshUnlessForced() throws IOExce
             }
         }
     }
+
+    public void testTypeWhenCompatible() throws IOException {
+        DocWriteResponse response = new DocWriteResponse(
+            new ShardId("index", "uuid", 0),
+            "id",
+            SequenceNumbers.UNASSIGNED_SEQ_NO,
+            17,
+            0,
+            DocWriteResponse.Result.CREATED
+        ) {
+            // DocWriteResponse is abstract so we have to sneak a subclass in here to test it.
+        };
+        try (XContentBuilder builder = JsonXContent.contentBuilder()) {
+            builder.withCompatibleVersion(RestApiVersion.V_7);
```

Comment:
I'm a bit uneasy that we can just change the compatible version on the XContentBuilder at any time, as that means that a consumer could change it accidentally; I think I'd prefer that it's a parameter passed to `contentBuilder()`.  But that's not really part of this issue, so we can look at it separately.

Reply:
I agree, these should be fixed, but also feel that this should be in a separate PR. The change might require touching many files and would be unrelated to the index/get api.
I will follow up in a separate PR

Reply:
also take a look at https://github.com/elastic/elasticsearch/pull/64423/files#r518884033
we try to prevent accidental change of the version with the `assert` inside `withCompatibleVersion`

Reply:
+1 to making it immutable or explicitly 'set once', that would clearly indicate the intent that it should never be updated.

Reply:
I refactored the XContentBuilder to have this field final and set in a constructor https://github.com/elastic/elasticsearch/pull/70878